integration-tests/src/main/java/com/datastax/oss/quarkus/tests/mapper/InventoryMapperProducer.java

```diff
@@ -19,45 +19,60 @@
+import java.util.concurrent.CompletionStage;
 import javax.enterprise.context.ApplicationScoped;
 import javax.enterprise.inject.Produces;
 import javax.inject.Inject;
 
 public class InventoryMapperProducer {
 
-  private final QuarkusCqlSession session;
-  private final InventoryMapper mapper;
+  private final CompletionStage<QuarkusCqlSession> sessionCompletionStage;
+  private final CompletionStage<InventoryMapper> mapperCompletionStage;
 
   @Inject
-  public InventoryMapperProducer(QuarkusCqlSession session) {
-    this.session = session;
-    mapper = new InventoryMapperBuilder(session).build();
+  public InventoryMapperProducer(CompletionStage<QuarkusCqlSession> sessionCompletionStage) {
+    this.sessionCompletionStage = sessionCompletionStage;
+    mapperCompletionStage =
+        sessionCompletionStage.thenApply(session -> new InventoryMapperBuilder(session).build());
   }
 
   @Produces
   @ApplicationScoped
-  ProductDao produceProductDao() {
-    return session
-        .getKeyspace()
-        .map(mapper::productDao)
-        .orElseThrow(() -> new IllegalStateException("Session is not bound to a keyspace"));
+  CompletionStage<ProductDao> produceProductDao() {
+    return sessionCompletionStage
+        .thenApply(
+            s ->
+                s.getKeyspace()
+                    .orElseThrow(
+                        () -> new IllegalStateException("Session is not bound to a keyspace")))
+        .thenCompose(
+            keyspace -> mapperCompletionStage.thenCompose(mapper -> mapper.productDao(keyspace)));
   }
 
   @Produces
   @ApplicationScoped
-  ProductReactiveDao produceProductReactiveDao() {
-    return session
-        .getKeyspace()
-        .map(mapper::productReactiveDao)
-        .orElseThrow(() -> new IllegalStateException("Session is not bound to a keyspace"));
+  CompletionStage<ProductReactiveDao> produceProductReactiveDao() {
+    return sessionCompletionStage
+        .thenApply(
+            s ->
+                s.getKeyspace()
+                    .orElseThrow(
+                        () -> new IllegalStateException("Session is not bound to a keyspace")))
+        .thenCompose(
+            keyspace ->
+                mapperCompletionStage.thenCompose(mapper -> mapper.productReactiveDao(keyspace)));
   }
 
   @Produces
   @ApplicationScoped
-  CustomerDao produceCustomerDao() {
-    return session
-        .getKeyspace()
-        .map(mapper::customerDao)
-        .orElseThrow(() -> new IllegalStateException("Session is not bound to a keyspace"));
+  CompletionStage<CustomerDao> produceCustomerDao() {
+    return sessionCompletionStage
+        .thenApply(
+            s ->
+                s.getKeyspace()
+                    .orElseThrow(
+                        () -> new IllegalStateException("Session is not bound to a keyspace")))
+        .thenCompose(
+            keyspace -> mapperCompletionStage.thenCompose(mapper -> mapper.customerDao(keyspace)));
```

Comment:
We've definitely lost some concision in the code here.  I was trying to get back to something closer to the original code with the following:

```java
-        .thenCompose(
-            keyspace -> mapperCompletionStage.thenCompose(mapper -> mapper.customerDao(keyspace)));
+        .thenCombine(mapperCompletionStage, (keyspace, mapper) -> mapper.customerDao(keyspace));
```

This fails because mapper.customerDao() now returns a ComplationStage<CustomerDao> rather than CustomerDao itself.  I mention this mainly in case anybody has a cleaner way to get at what I was trying to get at above... the nested thenCompose() calls are (primarily) what I was trying to avoid.

Reply:
As far as I know, the `thenCompose` pattern is recommended for achieving this goal using `CompletionStage` API